src/geometry/GeometryEvaluator.cc

```diff
@@ -73,34 +72,25 @@ std::shared_ptr<const Geometry> GeometryEvaluator::evaluateGeometry(const Abstra
     } else {
       this->traverse(node);
     }
-#ifdef ENABLE_CGAL
-    if (std::dynamic_pointer_cast<const CGALHybridPolyhedron>(this->root)) {
-      this->root = PolySetUtils::getGeometryAsPolySet(this->root);
-    }
-#endif
-#ifdef ENABLE_MANIFOLD
-    if (std::dynamic_pointer_cast<const ManifoldGeometry>(this->root)) {
-      this->root = PolySetUtils::getGeometryAsPolySet(this->root);
-    }
-#endif
```

Comment:
I believe the reason for doing this is that `GeometryCache` isn't really meant to cache any geometry. It's for caching PolySets (and Polygon2d). Otherwise, we wouldn't have a separate `CGALCache`.
..so my guess is that when extending with new geometry backends, it was easier to just convert to PolySet than to generalize the caching infrastructure.

Reply:
Ah, make sense. I think we need to rethink about this cache thing.

Reply:
I think the root cause triggering the STL issue is that `GeometryEvaluator::smartCacheInsert()` (https://github.com/openscad/openscad/blob/ded24c57b2eb91f6985697361bd435a31788941a/src/geometry/GeometryEvaluator.cc#L327) has buggy logic, causing it to insert `ManifoldGeometry` into the `GeometryCache`.

To fix that we need to rethink caching a bit.

Reply:
Yeah, I think we should try to make it clear what different parts of the API accepts. It would be cleaner if we can keep the backend specific types within the geometry part of the code, and use PolySet for everywhere else.